cpp/src/io/parquet/reader_impl_preprocess.cu

```diff
@@ -254,25 +269,57 @@ template <typename T = uint8_t>
       next_chunk++;
     }
     if (io_size != 0) {
-      auto& source = sources[chunk_source_map[chunk]];
-      if (source->is_device_read_preferred(io_size)) {
-        auto buffer        = rmm::device_buffer(io_size, stream);
-        auto fut_read_size = source->device_read_async(
-          io_offset, io_size, static_cast<uint8_t*>(buffer.data()), stream);
+      auto& source              = sources[chunk_source_map[chunk]];
+      auto const is_device_read = source->is_device_read_preferred(io_size);
+      reads.push_back({io_size, io_offset, is_compressed, is_device_read, chunk, next_chunk});
+      if (is_device_read) {
+        if (is_compressed) {
+          compressed_read_size += io_size;
+        } else {
+          uncompressed_read_size += io_size;
+        }
+      }
+    }
+    chunk = next_chunk;
+  }
+
+  auto uncompressed_buffer = rmm::device_buffer(uncompressed_read_size, stream);
+
+  auto compressed_buffer = rmm::device_buffer(compressed_read_size, stream);
+
+  auto compressed_buffer_data   = static_cast<uint8_t*>(compressed_buffer.data());
+  auto uncompressed_buffer_data = static_cast<uint8_t*>(uncompressed_buffer.data());
+  if (compressed_read_size > 0) {
+    page_data.emplace_back(datasource::buffer::create(std::move(compressed_buffer)));
+  }
+  if (uncompressed_read_size > 0) {
+    page_data.emplace_back(datasource::buffer::create(std::move(uncompressed_buffer)));
+  }
+
+  // queue reads
+  for (auto const& r : reads) {
+    auto d_compdata = [&]() -> const uint8_t* {
+      auto& source = sources[chunk_source_map[r.begin_chunk]];
+      if (r.device_read) {
+        auto const ptr     = r.compressed ? compressed_buffer_data : uncompressed_buffer_data;
+        auto fut_read_size = source->device_read_async(r.io_offset, r.io_size, ptr, stream);
         read_tasks.emplace_back(std::move(fut_read_size));
-        page_data[chunk] = datasource::buffer::create(std::move(buffer));
+        if (r.compressed) {
+          compressed_buffer_data += r.io_size;
+        } else {
+          uncompressed_buffer_data += r.io_size;
+        }
+        return ptr;
       } else {
-        auto const buffer = source->host_read(io_offset, io_size);
-        page_data[chunk] =
-          datasource::buffer::create(rmm::device_buffer(buffer->data(), buffer->size(), stream));
+        auto const buffer = source->host_read(r.io_offset, r.io_size);
+        page_data.emplace_back(
+          datasource::buffer::create(rmm::device_buffer(buffer->data(), buffer->size(), stream)));
+        return page_data.back()->data();
       }
-      auto d_compdata = page_data[chunk]->data();
-      do {
-        chunks[chunk].compressed_data = d_compdata;
-        d_compdata += chunks[chunk].compressed_size;
-      } while (++chunk != next_chunk);
-    } else {
-      chunk = next_chunk;
+    }();
+    for (auto chunk = r.begin_chunk; chunk < r.end_chunk; ++chunk) {
```

Comment:
Let's standardize the indexing variables with suffix `_id` so we can easily recognize it.
```suggestion
    for (auto chunk_id = r.begin_chunk; chunk_id < r.end_chunk; ++chunk_id) {
```